[Common Torizon] Layer refactor to accommodate Common Torizon distro and 3rd-party machines

EdTheBearded opened the pull request:
Initial refactor work of our layer to support the new `common-torizon` distro, and 2 new 3rd-party machine: x86 and NXP's iMX95, as well as a new Common Torizon distron for Xenomai, 3 and 4, on x86.

EdTheBearded commented:
Marking as draft as I'm still testing, but making the changes available for early review.

rborn-tx commented on conf/layer.conf:
Typo on the commit message title: "torzon" -> "torizon"

EdTheBearded replied:
done

rborn-tx commented on recipes-bsp/u-boot/u-boot-distro-boot.bb:
```diff
@@ -1,13 +1,22 @@
-FILESEXTRAPATHS:prepend := "${THISDIR}/${BPN}:"
+DESCRIPTION = "Boot script for launching images with U-Boot distro boot"
```
Nitpick: The other commit messages have the "Related-to" field set to "TOR-3661" while this commit says "Tor-3661".

EdTheBearded replied:
done

rborn-tx commented on recipes-images/images/torizon-base.inc:
```diff
@@ -24,7 +24,6 @@ SPLASH = "plymouth"
 
 # Base packages
```
Suggestion for the commit message: "recipes" -> "packages"

EdTheBearded replied:
done

rborn-tx reviewed (commented):
Awesome work!

I've raised a couple of questions just for clarification but I couldn't find any issues.

EdTheBearded commented on classes/image_type_common_torizon.bbclass:
```diff
@@ -0,0 +1,32 @@
+
+require torizon_base_image_type.inc
+
+EXTRA_OSTREE_COMMIT:remove = "--add-metadata=oe.layers="${OSTREE_LAYER_REVISION_INFO}""
```
Oh, good catch!

In fact we needed to remove this for iMX95, since we were having some build issues where we were passing `oe.layers=None`. And since it wasn't necessary I removed it. But I guess with all those changes it got global!

Let me add this line again and rerun some tests here. Thanks!!

EdTheBearded replied:
removed this line and everything is working fine.

EdTheBearded commented:
pushed some changes here:
 * @rborn-tx suggestion to disable provisioning data (https://github.com/torizon/meta-toradex-torizon/pull/202#discussion_r1901242219), so it's now disabled for Ccommon Torizon
 * added some configurations to imx95-19x19-verdin.inc, to disable some tasks while building Common Torizon
```
PREFERRED_PROVIDER_virtual/dtb = ""
OSTREE_DEPLOY_DEVICETREE:forcevariable = "0"
```
 * rebased this PR
 
 @jsrc27 this PR should be ready for review now. I'll keep it as a draft until the release is done and we're out of the freeze.

EdTheBearded commented on classes/image_type_common_torizon.bbclass:
```diff
@@ -0,0 +1,32 @@
+
+UBOOT_BINARY_OTA_IGNORE = "1"
+
+do_deploy[postfuncs] += "gen_torizon_prov_data"
```
done

jsrc27 reviewed (commented):
Just gave some minor feedback nothing too important. Otherwise changes look okay overall.

EdTheBearded commented on recipes-sota/grub-ota-fallback/grub-ota-fallback/99_fallback_logic:
```diff
@@ -0,0 +1,53 @@
+    save_env upgrade_available
+fi
+
+if [ "${bootcount}" == "${bootlimit}" ]; then
```
Thanks @jsrc27!
I've added a new commit on top, replacing this comparison for a `-ge` one, so that we avoid corner cases. Also notices some comparison with a single `=`, so I just fixed those in the same commit.

jsrc27 reviewed (dismissed):
Just preemptively giving my approval here.

EdTheBearded commented:
I've rebased the iMX95 commit to add a new Jira ticket it is related to: TOR-3712

EdTheBearded commented on classes/image_type_torizon.bbclass:
```diff
@@ -27,620 +27,29 @@ python adjust_tezi_artifacts() {
 UBOOT_BINARY_OTA:colibri-imx6 = "u-boot-with-spl.imx"
 UBOOT_BINARY_OTA:colibri-imx6ull-emmc = "u-boot.imx"
 UBOOT_BINARY_OTA:colibri-imx7-emmc = "u-boot.imx"
 UBOOT_BINARY_OTA:apalis-imx8 = "imx-boot"
```
Removed this `UBOOT_BINARY_OTA:*` that was being repeated on `torizon_base_image_type.inc`. So now we have these defined only once and where they're relevant

EdTheBearded commented:
@jsrc27 and @rborn-tx, I've pointed the changes I made on the above comments. Other than that, it was just a rebase (hopefully the last one).

This PR is ready to be reviewed and merged!

jsrc27 reviewed (dismissed):
The additional changes LGTM.

microhobby commented on conf/distro/common-torizon-xenomai4.conf:
```diff
@@ -0,0 +1,11 @@
+require conf/distro/torizon-common.conf
```
it should not be `common-torizon.conf` ?

microhobby commented on conf/distro/common-torizon-xenomai3.conf:
```diff
@@ -0,0 +1,11 @@
+require conf/distro/torizon-common.conf
```
it should not be `common-torizon.conf` ?

microhobby commented on conf/distro/common-torizon.conf:
```diff
@@ -0,0 +1,7 @@
+require conf/distro/sota.conf.inc
+
+DISTRO = "common-torizon"
+DISTRO_NAME = "Common Torizon OS"
```
I think that we should follow the same pattern from upstream something like "Torizon OS Common" instead.

microhobby replied:
@EdTheBearded this does not seem to be resolved. At least one answer like "we decided to follow this decision page ...", "the product manager ask me to follow with this ..." or "we need to decide ..." instead of simple resolve without an answer.

@leograba I did my review and I was ignored for a lot of points. So, next time if someone asks why I'm not involved and would not like to contribute you know why.

microhobby commented on classes/torizon_base_image_type.inc:
```diff
@@ -0,0 +1,594 @@
+    import subprocess
+    try:
+        layers = []
+        paths = (d.getVar("BBLAYERS" or "")).split()
```
any reason by sometimes use `d.getVar("VAR")` and others `d.getVar('VAR')`?

microhobby commented on classes/torizon_base_image_type.inc:
```diff
@@ -0,0 +1,594 @@
+            rev, _ = bb.process.run('export PSEUDO_UNLOAD=1; git rev-parse HEAD', cwd=path)
+            branch, _ = bb.process.run('export PSEUDO_UNLOAD=1; git rev-parse --abbrev-ref HEAD', cwd=path)
+            try:
+                subprocess.check_output("""cd %s; export PSEUDO_UNLOAD=1; set -e;
```
why not use `b.process.run` instead of `subprocess.check_output`?

rborn-tx reviewed (dismissed):
LGTM

EdTheBearded commented:
Resolved another rebase.
Also changed recipes-kernel/linux/linux-torizon.inc to not append anymore `${MACHINE}.cfg`, since this is ssmething we only used for x86, and all other machines would require an empty .cfg file for the build to conclude.
Nout, on x86 specific kernel bbappends, we appends the required fragments.

EdTheBearded commented:
@rborn-tx and @jsrc27 could you guys please review this again, so we can merge this before we need yet another rebase?
